hostname is not set if dhcp is used for assigning addresses

##  System Information

### Linux distribution

``` opensuse leap 15 sp1```

### Terraform version
```
Terraform v0.12.19
+ provider.libvirt (unversioned
```

### Provider and libvirt versions

```
terraform-provider-libvirt was not built correctly
Compiled against library: libvirt 5.1.0
Using library: libvirt 5.1.0
```
Note:  the provider was installed from package `0.6.1+git.1580485189.cc55d5d4-22.1` 

## Checklist

- [ ] Is your issue/contribution related with enabling some setting/option exposed by libvirt that the plugin does not yet support, or requires changing/extending the provider terraform schema?
- [ ] Make sure you explain why this option is important to you, why it should be important to everyone. Describe your use-case with detail and provide examples where possible.
- [ ] If it is a very special case, consider using the _XSLT_ support in the provider to tweak the definition instead of opening an issue
- [ ] Maintainers do not have expertise in every libvirt setting, so please, describe the feature and how it is used. Link to the appropriate documentation
- [X] Is it a bug or something that does not work as expected? Please make sure you fill the version information below:

## Description of Issue/Question

When assigning ip addresses dynamically from a nat network, the hostname specified in the network interface is not respected. If the address is assigned statically, it is respected.

### Setup
```
provider "libvirt" {
  uri = "qemu://system"
}

resource "libvirt_network" "network" {
  name   = "test-network"
  mode   = "nat"
  domain = "test.local"

  dns {
    enabled = true
  }

  addresses = ["10.17.0.0/22"]
}

resource "libvirt_volume" "img" {
  name   = "SLE-15-SP1-JeOS-GMC.qcow2"
  source = "path/to/image"
  pool   = "default"
}
resource "libvirt_volume" "master" {
  name           = "master-volume"
  pool           = "default"
  size           = "25769803776"
  base_volume_id = libvirt_volume.img.id
}

resource "libvirt_domain" "master" {
  name       = "master-domain"

  disk {
    volume_id = libvirt_volume.master.id
  }

  network_interface {
    network_id     = libvirt_network.network.id
    hostname       = "master"
    wait_for_lease = true
  }

  graphics {
    type        = "vnc"
    listen_type = "address"
  }
}
```


### Steps to Reproduce Issue

Apply the configuration as above and the resulting `libvirt_domain` resource will have the  network interface defined as shown below:
```
            "network_interface": [
              {
                "addresses": [
                  "10.17.0.198"
                ],
                "hostname": "",
```

If the configuration is changed to add a fixed ip address, 
```
  network_interface {
    network_id     = libvirt_network.network.id
    hostname      =  "master"
   addresses      =  [ "10.17.1.20"] 
   wait_for_lease = true
  }
```

the hostname is respected.

```
            "network_interface": [
              {
                "addresses": [
                  "10.17.1.20"
                ],
                "hostname": "master",
```


___
## Additional information:

Do you have SELinux or Apparmor/Firewall enabled? Some special configuration?
Have you tried to reproduce the issue without them enabled?

I have modified the acceptance test to check the hostname and not they fail
```
$ go test --run TestAccLibvirtDomain_NetworkInterface -v
=== RUN   TestAccLibvirtDomain_NetworkInterface
--- FAIL: TestAccLibvirtDomain_NetworkInterface (21.26s)
    testing.go:635: Step 0 error: Check failed: Check 4/4 error: libvirt_domain.cwbekjwsrk: Attribute 'network_interface.1.hostname' expected "myhost", got ""
FAIL
exit status 1
```


This is the patch to the tests. 
```
diff --git a/libvirt/resource_libvirt_domain_test.go b/libvirt/resource_libvirt_domain_test.go
index 8ed1fd27..ba1569b5 100644
--- a/libvirt/resource_libvirt_domain_test.go
+++ b/libvirt/resource_libvirt_domain_test.go
@@ -607,6 +607,7 @@ func TestAccLibvirtDomain_NetworkInterface(t *testing.T) {
                }
                network_interface  {
                        network_name   = "default"
+                        hostname       = "myhost"
                        mac            = "52:54:00:A9:F5:17"
                        wait_for_lease = true
                }
@@ -622,12 +623,15 @@ func TestAccLibvirtDomain_NetworkInterface(t *testing.T) {
                Steps: []resource.TestStep{
                        {
                                Config: config,
+                                ExpectNonEmptyPlan: true,
                                Check: resource.ComposeTestCheckFunc(
                                        testAccCheckLibvirtDomainExists("libvirt_domain."+randomDomainName, &domain),
                                        resource.TestCheckResourceAttr(
                                                "libvirt_domain."+randomDomainName, "network_interface.0.network_name", "default"),
                                        resource.TestCheckResourceAttr(
                                                "libvirt_domain."+randomDomainName, "network_interface.1.mac", "52:54:00:A9:F5:17"),
+                                       resource.TestCheckResourceAttr(
+                                               "libvirt_domain."+randomDomainName, "network_interface.1.hostname", "myhost"),
                                ),
                        },
                },
``` 

Notice that I have to add the `ExpectNonEmptyPlan` option because otherwise, it fails with this message:
```
$go test --run TestAccLibvirtDomain_NetworkInterface -v
=== RUN   TestAccLibvirtDomain_NetworkInterface
--- FAIL: TestAccLibvirtDomain_NetworkInterface (11.08s)
    testing.go:635: Step 0 error: After applying this step, the plan was not empty:
```
Followed by a large diff that includes this line 
```
          network_interface.1.hostname:       "" => "myhost"
``` 

This is consistent with the warning messages from terraform: 
```
2020/02/25 09:27:32 [WARN] Provider "registry.terraform.io/-/libvirt" produced an unexpected new value for libvirt_domain.master, but we are tolerating it because it is using the legacy plugin SDK.
```

--- a/libvirt/domain.go
+++ b/libvirt/domain.go
@@ -706,23 +706,58 @@ func setNetworkInterfaces(d *schema.ResourceData, domainDef *libvirtxml.Domain,
 		}
 
 		// connect to the interface to the network... first, look for the network
-		if n, ok := d.GetOk(prefix + ".network_name"); ok {
-			// when using a "network_name" we do not try to do anything: we just
-			// connect to that network
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Network: &libvirtxml.DomainInterfaceSourceNetwork{
-					Network: n.(string),
-				},
+		var network *libvirt.Network
+		var err error
+
+		if networkName, ok := d.GetOk(prefix + ".network_name"); ok {
+			network, err = virConn.LookupNetworkByName(networkName.(string))
+			if err != nil {
+				return fmt.Errorf("Can't retrieve network '%s'", networkName.(string))
 			}
+			defer network.Free()
+
 		} else if networkUUID, ok := d.GetOk(prefix + ".network_id"); ok {
 			// when using a "network_id" we are referring to a "network resource"
 			// we have defined somewhere else...
-			network, err := virConn.LookupNetworkByUUIDString(networkUUID.(string))
+			network, err = virConn.LookupNetworkByUUIDString(networkUUID.(string))
 			if err != nil {
 				return fmt.Errorf("Can't retrieve network ID %s", networkUUID)
 			}
 			defer network.Free()
 
+		} else if bridgeNameI, ok := d.GetOk(prefix + ".bridge"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Bridge: &libvirtxml.DomainInterfaceSourceBridge{
+					Bridge: bridgeNameI.(string),
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".vepa"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "vepa",
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".macvtap"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "bridge",
+				},
+			}
+		} else if devI, ok := d.GetOk(prefix + ".passthrough"); ok {
+			netIface.Source = &libvirtxml.DomainInterfaceSource{
+				Direct: &libvirtxml.DomainInterfaceSourceDirect{
+					Dev:  devI.(string),
+					Mode: "passthrough",
+				},
+			}
+		} else {
+			// no network has been specified: we are on our own
+		}
+
+		// if we got a network
+		if network != nil {
 			networkName, err := network.GetName()
 			if err != nil {
 				return fmt.Errorf("Error retrieving network name: %s", err)
@@ -763,10 +798,13 @@ func setNetworkInterfaces(d *schema.ResourceData, domainDef *libvirtxml.Domain,
 						// have a valid lease and then read the IP we have been assigned, so we can
 						// do the mapping
 						log.Printf("[DEBUG] Do not have an IP for '%s' yet: will wait until DHCP provides one...", hostname)
+						if err != nil {
+							return err
+						}
 						partialNetIfaces[strings.ToUpper(mac)] = &pendingMapping{
-							mac:      strings.ToUpper(mac),
-							hostname: hostname,
-							network:  network,
+							mac:         strings.ToUpper(mac),
+							hostname:    hostname,
+							networkName: networkName,
 						}
 					}
 				}
@@ -777,35 +815,6 @@ func setNetworkInterfaces(d *schema.ResourceData, domainDef *libvirtxml.Domain,
 					Network: networkName,
 				},
 			}
-		} else if bridgeNameI, ok := d.GetOk(prefix + ".bridge"); ok {
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Bridge: &libvirtxml.DomainInterfaceSourceBridge{
-					Bridge: bridgeNameI.(string),
-				},
-			}
-		} else if devI, ok := d.GetOk(prefix + ".vepa"); ok {
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Direct: &libvirtxml.DomainInterfaceSourceDirect{
-					Dev:  devI.(string),
-					Mode: "vepa",
-				},
-			}
-		} else if devI, ok := d.GetOk(prefix + ".macvtap"); ok {
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Direct: &libvirtxml.DomainInterfaceSourceDirect{
-					Dev:  devI.(string),
-					Mode: "bridge",
-				},
-			}
-		} else if devI, ok := d.GetOk(prefix + ".passthrough"); ok {
-			netIface.Source = &libvirtxml.DomainInterfaceSource{
-				Direct: &libvirtxml.DomainInterfaceSourceDirect{
-					Dev:  devI.(string),
-					Mode: "passthrough",
-				},
-			}
-		} else {
-			// no network has been specified: we are on our own
 		}
 
 		domainDef.Devices.Interfaces = append(domainDef.Devices.Interfaces, netIface)

--- a/libvirt/resource_libvirt_domain.go
+++ b/libvirt/resource_libvirt_domain.go
@@ -17,9 +17,9 @@ import (
 )
 
 type pendingMapping struct {
-	mac      string
-	hostname string
-	network  *libvirt.Network
+	mac         string
+	hostname    string
+	networkName string
 }
 
 func init() {
@@ -144,9 +144,9 @@ func resourceLibvirtDomain() *schema.Resource {
 				},
 			},
 			"disk": {
-				Type:     schema.TypeList,
-				Optional: true,
-				ForceNew: true,
+				Type:       schema.TypeList,
+				Optional:   true,
+				ForceNew:   true,
 				ConfigMode: schema.SchemaConfigModeAttr,
 				Elem: &schema.Resource{
 					Schema: map[string]*schema.Schema{
@@ -586,7 +586,6 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 	// we must read devices again in order to set some missing ip/MAC/host mappings
 	for i := 0; i < d.Get("network_interface.#").(int); i++ {
 		prefix := fmt.Sprintf("network_interface.%d", i)
-
 		mac := strings.ToUpper(d.Get(prefix + ".mac").(string))
 
 		// if we were waiting for an IP address for this MAC, go ahead.
@@ -594,16 +593,26 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 			// we should have the address now
 			addressesI, ok := d.GetOk(prefix + ".addresses")
 			if !ok {
-				return fmt.Errorf("Did not obtain the IP address for MAC=%s", mac)
+				log.Printf("Did not obtain the IP address for MAC=%s", mac)
+				continue
 			}
+
+			network, err := virConn.LookupNetworkByName(pending.networkName)
+			if err != nil {
+				log.Printf("Can't retrieve network '%s'", pending.networkName)
+				continue
+			}
+
 			for _, addressI := range addressesI.([]interface{}) {
 				address := addressI.(string)
 				log.Printf("[INFO] Finally adding IP/MAC/host=%s/%s/%s", address, mac, pending.hostname)
-				updateOrAddHost(pending.network, address, mac, pending.hostname)
+
+				err = updateOrAddHost(network, address, mac, pending.hostname)
 				if err != nil {
-					return fmt.Errorf("Could not add IP/MAC/host=%s/%s/%s: %s", address, mac, pending.hostname, err)
+					log.Printf("Could not add IP/MAC/host=%s/%s/%s: %s", address, mac, pending.hostname, err)
 				}
 			}
+			network.Free()
 		}
 	}
 
@@ -914,6 +923,7 @@ func resourceLibvirtDomainRead(d *schema.ResourceData, meta interface{}) error {
 		}
 
 		netIface["wait_for_lease"] = d.Get(prefix + ".wait_for_lease").(bool)
+		netIface["hostname"] = d.Get(prefix + ".hostname").(string)
 		netIface["addresses"] = addressesForMac(mac)
 		log.Printf("[DEBUG] read: addresses for '%s': %+v", mac, netIface["addresses"])
 

--- a/libvirt/resource_libvirt_domain_test.go
+++ b/libvirt/resource_libvirt_domain_test.go
@@ -66,7 +66,7 @@ func TestAccLibvirtDomain_Description(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"libvirt_domain."+randomResourceName, "name", randomDomainName),
 					resource.TestCheckResourceAttr(
-                        "libvirt_domain."+randomResourceName, "description", "unit test description"),
+						"libvirt_domain."+randomResourceName, "description", "unit test description"),
 				),
 			},
 		},
@@ -603,17 +603,21 @@ func TestAccLibvirtDomain_NetworkInterface(t *testing.T) {
 	resource "libvirt_domain" "%s" {
 		name              = "%s"
 		network_interface  {
-			network_name = "default"
+                        network_name = "${libvirt_network.%s.name}"
+                        hostname       = "myhost"
+			mac            = "52:54:00:A9:F5:17"
+			wait_for_lease = true
 		}
 		network_interface  {
-			network_name   = "default"
-			mac            = "52:54:00:A9:F5:17"
+                        network_id = "${libvirt_network.%s.id}"
+                        hostname       = "myhost"
+			mac            = "52:54:00:A9:F5:19"
 			wait_for_lease = true
 		}
 		disk {
 			file = "%s/testdata/tcl.iso"
 		}
-	}`, randomNetworkName, randomNetworkName, randomDomainName, randomDomainName, currentDir)
+	}`, randomNetworkName, randomNetworkName, randomDomainName, randomDomainName, randomNetworkName, randomNetworkName, currentDir)
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -621,13 +625,23 @@ func TestAccLibvirtDomain_NetworkInterface(t *testing.T) {
 		CheckDestroy: testAccCheckLibvirtDomainDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: config,
+				Config:             config,
+				ExpectNonEmptyPlan: false, // TODO: find why if set to true, the test fails
+				Destroy:            false,
 				Check: resource.ComposeTestCheckFunc(
 					testAccCheckLibvirtDomainExists("libvirt_domain."+randomDomainName, &domain),
 					resource.TestCheckResourceAttr(
-						"libvirt_domain."+randomDomainName, "network_interface.0.network_name", "default"),
+						"libvirt_domain."+randomDomainName, "network_interface.0.network_name", randomNetworkName),
 					resource.TestCheckResourceAttr(
-						"libvirt_domain."+randomDomainName, "network_interface.1.mac", "52:54:00:A9:F5:17"),
+						"libvirt_domain."+randomDomainName, "network_interface.0.mac", "52:54:00:A9:F5:17"),
+					resource.TestCheckResourceAttr(
+						"libvirt_domain."+randomDomainName, "network_interface.0.hostname", "myhost"),
+					resource.TestCheckResourceAttr(
+						"libvirt_domain."+randomDomainName, "network_interface.1.network_name", randomNetworkName),
+					resource.TestCheckResourceAttr(
+						"libvirt_domain."+randomDomainName, "network_interface.1.mac", "52:54:00:A9:F5:19"),
+					resource.TestCheckResourceAttr(
+						"libvirt_domain."+randomDomainName, "network_interface.1.hostname", "myhost"),
 				),
 			},
 		},

